### PR TITLE
En 4132 make every account in trie unique

### PR DESCRIPTION
**Pull request by BeniaminDrasovean**

### CODE REVIEW

File: data/interface.go [hunk 1]
```diff
@@ -94,8 +94,10 @@ type Trie interface {
 	Recreate(root []byte) (Trie, error)
 	String() string
 	DeepClone() (Trie, error)
-	Rollback(rootHash []byte) error
+	Rollback(rootHash []byte)
 	Prune(rootHash []byte) error
+	ResetOldHashes() [][]byte
+	AddHashesToOldHashes([][]byte)
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply data/interface.go]: Renamed. Sounds a lot better.

File: data/state/accountsDB.go [hunk 2]
```diff
@@ -494,6 +501,16 @@ func (adb *AccountsDB) clearJournal() {
 	adb.mutEntries.Unlock()
 }
 
+// PruneTrie removes old values from the trie database
+func (adb *AccountsDB) PruneTrie(rootHash []byte) error {
+	return adb.mainTrie.Prune(rootHash)
+}
+
+// Rollback clears the trie's evictionWaitingList
+func (adb *AccountsDB) Rollback(rootHash []byte) {
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply data/state/accountsDB.go]: Renamed.

File: data/state/trackableDataTrie_test.go [hunk 3]
```diff
@@ -23,58 +23,78 @@ func TestTrackableDataAccountRetrieveValueNilDataTrieShouldErr(t *testing.T) {
...
+	identifier := []byte("identifier")
 	trie := &mock.TrieStub{}
-	tdaw := state.NewTrackableDataTrie(trie)
+	tdaw := state.NewTrackableDataTrie(identifier, trie)
 	assert.NotNil(t, tdaw)
 
 	tdaw.SetDataTrie(&mock.TrieStub{})
-	tdaw.DirtyData()["ABC"] = []byte{32, 33, 34}
+	key := []byte("ABC")
+	val := []byte{32, 33, 34}
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply data/state/trackableDataTrie_test.go]: Done.

File: integrationTests/state/stateTrie_test.go [hunk 4]
```diff
@@ -737,7 +737,7 @@ func TestAccountsDB_RevertDataStepByStepWithCommitsAccountDataShouldWork(t *test
 	rootHash, err = adb.RootHash()
 	assert.Nil(t, err)
 	hrCreated2 := base64.StdEncoding.EncodeToString(rootHash)
-	rootHash, err = state1.DataTrie().Root()
+	rootHash, err = state2.DataTrie().Root()
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply integrationTests/state/stateTrie_test.go]: Done.

File: integrationTests/state/stateTrie_test.go [hunk 5]
```diff
@@ -749,7 +749,7 @@ func TestAccountsDB_RevertDataStepByStepWithCommitsAccountDataShouldWork(t *test
 	assert.NotEqual(t, hrCreated1, hrCreated2)
 
 	//Test 2.2 test whether the datatrie roots match
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply integrationTests/state/stateTrie_test.go]: Updated.

File: process/sync/shardblock.go [hunk 6]
```diff
@@ -898,6 +898,8 @@ func (boot *ShardBootstrap) rollback(header *block.Header) error {
 	var newRootHash []byte
 
 	if header.GetNonce() > 1 {
+		boot.accounts.Rollback(header.GetRootHash())
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply process/sync/shardblock.go]: It works correctly. Rename, so the purpose of this call is more obvious.

[the remark this answers is not in the document]
Comment (iulianpascalau) [reply process/sync/shardblock.go]: Ok, CancelPrune sounds better :+1:

File: integrationTests/state/stateTrie_test.go [hunk 7]
```diff
@@ -1069,3 +1068,96 @@ func BenchmarkTxExecution(b *testing.B) {
 		integrationTests.AdbEmulateBalanceTxSafeExecution(acntSrc.(*state.Account), acntDest.(*state.Account), adb, big.NewInt(1))
 	}
 }
+
```
Comment (AdoAdoAdo) [inline integrationTests/state/stateTrie_test.go]: check warnings on file, shadowing variable err can be fixed.

File: process/block/shardblock.go [hunk 8]
```diff
@@ -733,6 +733,18 @@ func (sp *shardProcessor) CommitBlock(
 		log.Debug(errNotCritical.Error())
 	}
 
+	for i := range finalHeaders {
+		rootHash := finalHeaders[i].GetRootHash()
+		if rootHash == nil {
+			continue
+		}
+
+		errNotCritical := sp.accounts.PruneTrie(rootHash)
```
Comment (AdoAdoAdo) [inline process/block/shardblock.go]: Can we add a test for the scenario where:
1. We have one block "BlockX" with a new root hash
2. A number of blocks are produced with no transactions (same root hash) 
3. "BlockX" becomes final so Prune is called

Expected results:
Root hash from block X is still in the trie, the Pruning does not remove it with its value

File: process/sync/shardblock.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (AdoAdoAdo) [reply process/sync/shardblock.go]: here I think you should "Rollback" aka cancel pruning for the previous root hash not the current block root hash, as this is the last committed block root hash and was not added for eviction.

File: data/state/trackableDataTrie_test.go [hunk 9]
```diff
@@ -23,58 +23,78 @@ func TestTrackableDataAccountRetrieveValueNilDataTrieShouldErr(t *testing.T) {
...
 
+	identifier := []byte("identifier")
 	trie := &mock.TrieStub{}
-	tdaw := state.NewTrackableDataTrie(trie)
+	tdaw := state.NewTrackableDataTrie(identifier, trie)
 	assert.NotNil(t, tdaw)
 
 	tdaw.SetDataTrie(&mock.TrieStub{})
-	tdaw.DirtyData()["ABC"] = []byte{32, 33, 34}
+	key := []byte("ABC")
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply data/state/trackableDataTrie_test.go]: Done.

File: data/trie/node_test.go [hunk 10]
```diff
@@ -581,3 +581,62 @@ func TestTrieDatabasePruning(t *testing.T) {
 		assert.NotNil(t, err)
 	}
 }
+
+func TestTrieResetOldHashes(t *testing.T) {
+	t.Parallel()
+
+	msh, hsh := getTestMarshAndHasher()
+	evictionCacheSize := 100
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply data/trie/node_test.go]: Done. Renamed everywhere.

File: integrationTests/state/stateTrie_test.go (hunk 7 above)
Comment (BeniaminDrasovean) [reply integrationTests/state/stateTrie_test.go]: Done.

File: process/block/shardblock.go (hunk 8 above)
Comment (BeniaminDrasovean) [reply process/block/shardblock.go]: Will add a todo, and will open another pull request with integration tests.

File: process/sync/shardblock.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply process/sync/shardblock.go]: Right, changed. Also added a todo to create unit tests to verify that pruning and pruning canceling work properly.

File: process/block/shardblock.go [hunk 11]
```diff
@@ -739,7 +739,8 @@ func (sp *shardProcessor) CommitBlock(
 			continue
 		}
 
-		errNotCritical := sp.accounts.PruneTrie(rootHash)
+		// TODO add integration test to see that pruning works as expected
+		errNotCritical = sp.accounts.PruneTrie(rootHash)
```
Comment (AdoAdoAdo) [inline process/block/shardblock.go]: I suspect that you have to change to prevBlockHeader(finalHeaders[i]).GetRootHash() here

Comment (BeniaminDrasovean) [reply process/block/shardblock.go]: Done.

File: process/block/shardblock.go [hunk 12]
```diff
@@ -733,8 +733,21 @@ func (sp *shardProcessor) CommitBlock(
 		log.Debug(errNotCritical.Error())
 	}
 
+	headersStorage := sp.dataPool.Headers()
 	for i := range finalHeaders {
-		rootHash := finalHeaders[i].GetRootHash()
+		val, ok := headersStorage.Get(finalHeaders[i].GetPrevHash())
+		if !ok {
+			log.Debug("Could not get header from storage")
```
Comment (AdoAdoAdo) [inline process/block/shardblock.go]: lower case string?

Comment (BeniaminDrasovean) [reply process/block/shardblock.go]: done

File: process/block/shardblock.go (hunk 11 above)
Comment (AdoAdoAdo) [reply process/block/shardblock.go]: need to take it from storage instead of pool

**Pull request merged**